### ISSUE TITLE
lint “non_upper_case_globals” wrongly warns the consumer

### ISSUE THREAD

**Issue by nwin** [rust-lang/rust#25207]
The lint “non_upper_case_globals” wrongly warns the consumer about a wrongly named constant ([playpen](http://is.gd/GplrQb)). It only seems to happen if the constant is imported into the current namespace and if you’re using it in a match statement:

```rust
mod consts {
    #![allow(non_upper_case_globals)]
    pub const fooBAR: usize = 23;
}

fn main() {
    use consts::*;
    println!("{}", match 23 {
        fooBAR => "yes",
        _ => "no"
    });
}
```

**Comment by pnkfelix** [comment]
I think this is by design; someone reading the match arm `fooBAR => ...` is likely to interpret `fooBAR` as a binding, not a constant that is being matched against...

**Comment by nwin** [comment]
I don’t think so. You have the same issue when matching against lower-case enum variants and for this is no warning at the match site.

**Comment by pnkfelix** [comment]
I think I can agree that there is an "if-and-only-if" relationship here: if (and only if) we warn about `fooBAR => ...` when `fooBAR` is a `const`-item, then we should warn about `fooBAR => ...` when `fooBAR` is an `enum`-variant.

**Comment by nwin** [comment]
I think there are two things to agree on:

1. What is the purpose of `non_upper_case_globals`? Warning about naming conventions or ambiguity?
2. Do we want to warn about ambiguity (because of expectations) in this case?

If it is the purpose of  `non_upper_case_globals` to warning about naming conventions this lint should not warn about this case.

On the other hand, if it’s purpose is to warn about ambiguity, and we want to warn about ambiguity in this case it should give a better hint (e.g. "use full qualifier").

Furthermore we would also have to warn about this case (because it is *not* a constant):

```rust
fn main() {
    println!("{}", match 23 {
        FOO_BAR => "yes",
        _ => "no"
    });
}
```

This is why I oppose. The consequence would would be that would also have to warn about every variable not following naming conventions.

**Comment by steveklabnik** [comment]
Triage: no changes

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by Lokathor** [comment]
This seems to have been filed a second time: https://github.com/rust-lang/rust/issues/39371

I presume one of the two issues can be merged into the other.

**Comment by Spoonbender** [comment]
Triage: no change

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Nominating this on behalf of @Lokathor.  The main issue here seems to be that the person whose code is getting the warning is not necessarily well positioned to change the name, and the upstream who is well placed to change the name may have already silenced this warning but we still warn in the downstream code.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting.

We agreed that the current lint should not fire on uses, only on declarations. However, we also felt cases that are ambiguous between being a constant or a binding (e.g. bare names like `xyzABC` without a `::`), we should have a *different* lint to flag those and suggest that they be made unambiguously either a binding or a constant.

A pattern could be made unambiguously a binding with an `@_`, or made unambiguously a constant by giving a path that includes a `::`.
